prov/gni/test/fi_addr_str.c

```diff
@@ -0,0 +1,1660 @@
+
+		cr_assert(sz == FI_SUCCESS, "Invalid return value: %s",
+			  fi_strerror((int) -sz));
+
+		d_expected_cqe.buf = NULL;
+		d_expected_cqe.data = 0;
+		d_expected_cqe.flags = (FI_MSG | FI_RECV);
+		d_expected_cqe.len = len;
+		d_expected_cqe.op_context = source[i];
+		d_expected_cqe.tag = 0;
+
+		sz = fi_recv(rx_ep[NUMEPS - 1 - i][idx],
+			     target[NUMEPS - 1 - i], len,
+			     NULL, gni_addr[i], source[i]);
+
+		cr_assert(sz == FI_SUCCESS, "Invalid return value: %s",
+			  fi_strerror((int) -sz));
+
+		src_done = dest_done = 0;
+		/* Progress sender and receiver */
+		do {
+			ret = fi_cq_read(tx_cq[i][idx], &s_cqe, 1);
+			if (ret == 1)
+				src_done = 1;
+
+			ret = fi_cq_read(rx_cq[NUMEPS - 1 - i][idx],
+					 &d_cqe, 1);
+			if (ret == 1)
+				dest_done = 1;
+		} while (src_done != 1 || dest_done != 1);
+
+		cntr = fi_cntr_read(send_cntr[i]);
+		cr_assert(cntr == ++sends[i],
+			  "Invalid send counter: actual(%lu), expected(%lu)",
+			  cntr, sends[i]);
+
+		cntr = fi_cntr_read(recv_cntr[NUMEPS - 1 - i]);
+		cr_assert(cntr == ++recvs[NUMEPS - 1 - i],
+			  "Invalid recv counter: actual(%lu), expected(%lu)",
+			  cntr, recvs[NUMEPS - 1 - i]);
+
+		check_tagged_cqe(s_expected_cqe, s_cqe);
+		check_tagged_cqe(d_expected_cqe, d_cqe);
+
+		check_buf(source[i], target[NUMEPS - 1 - i], len);
+	}
+}
+
+void do_send_recv(void)
+{
+	int len, i, j;
+
+	switch (ep_type) {
+	case EP:
+		for (len = 2; len <= BUF_SZ; len *= 2) {
+			do_ep_send_recv_iter(len);
+		}
+		break;
+
+	case SEP:
+		for (j = 0; j < ctx_cnt; j++) {
+			for (len = 2; len <= BUF_SZ; len *= 2) {
+				do_sep_send_recv_iter(j, len);
+			}
+
+			for (i = 0; i < NUMEPS; i++) {
+				fi_cntr_set(send_cntr[i], 0);
+				fi_cntr_set(recv_cntr[i], 0);
+			}
+		}
+		break;
+	case PEP:
+		break;
+
+	default:
+		cr_assert_fail("Invalid endpoint type.");
+	}
+}
+
+void do_invalid_fi_getinfo(void)
+{
+	int i, ret;
+
+	/* Get info about fabric services with the provided hints */
+	for (i = 0; i < NUMEPS; i++) {
+		ret = fi_getinfo(FI_VERSION(1, 0), "this is a test", "testing",
+				 0, hints, &fi[i]);
+		cr_assert(ret == -FI_ENODATA, "fi_getinfo returned: %s",
+			  fi_strerror(-ret));
+
+		fi_freeinfo(fi[i]);
+	}
+}
+
+void do_valid_fi_getinfo_with_fas(void)
+{
+	int i, ret;
+	char *fas = calloc(GNIX_FI_ADDR_STR_LEN, sizeof(char));
+	struct gnix_ep_name ep_name;
+
+	/* Test without flags */
+	for (i = 0; i < NUMEPS; i++) {
+		generate_rand_fas(&fas);
+
+		ret = fi_getinfo(FI_VERSION(1, 0), fas, NULL, 0, hints, &fi[i]);
+		cr_assert(ret == FI_SUCCESS, "fi_getinfo returned: %s",
+			  fi_strerror(-ret));
+
+		dbg_printf(BLUE "fi[%d]->dest_addr = %s\n" COLOR_RESET,
+			   i, (char *) fi[i]->dest_addr);
+
+		fas_to_ep_name(fas, &ep_name);
+		check_ep_name(((struct gnix_ep_name *) fi[i]->dest_addr)[0],
+			      ep_name);
+
+		fi_freeinfo(fi[i]);
+	}
+
+	/* Test with FI_SOURCE */
+	for (i = 0; i < NUMEPS; i++) {
+		generate_rand_fas(&fas);
```

Comment:
Lines 1372-1374 below verify the matching of this data. fi_getinfo results in the node parameter being parsed and converted to a `struct gnix_ep_name` -- the result of this is stored in `fi[i]->src_addr`. We verify that this `fi[i]->src_addr` matches the result of converting the random string, `fas`, and storing the result of that conversion in `ep_name`. Finally, we compare the structure members of `ep_name` to `fi[i]->src_addr`.